pkg/embeddedcluster/helmvm_node.go

```diff
@@ -0,0 +1,168 @@
+	if err == nil {
+		if nodeUsageMetrics.Usage.Memory() != nil {
+			memoryCapacity.Used = float64(nodeUsageMetrics.Usage.Memory().Value()) / math.Pow(2, 30)
+		}
+
+		if nodeUsageMetrics.Usage.Cpu() != nil {
+			cpuCapacity.Used = nodeUsageMetrics.Usage.Cpu().AsApproximateFloat64()
+		}
+	} else {
+		// if we can't get metrics, we'll do nothing for now
+		// in the future we may decide to retry or log a warning
+	}
+
+	podCapacity.Used = float64(len(nodePods))
+
+	podInfo := []types.PodInfo{}
+
+	for _, pod := range nodePods {
+		newInfo := types.PodInfo{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+			Status:    string(pod.Status.Phase),
+		}
+
+		podMetrics, err := metricsClient.MetricsV1beta1().PodMetricses(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err == nil {
+			podTotalMemory := 0.0
+			podTotalCPU := 0.0
+			for _, container := range podMetrics.Containers {
+				if container.Usage.Memory() != nil {
+					podTotalMemory += float64(container.Usage.Memory().Value()) / math.Pow(2, 20)
+				}
+				if container.Usage.Cpu() != nil {
+					podTotalCPU += container.Usage.Cpu().AsApproximateFloat64() * 1000
+				}
+			}
+			newInfo.Memory = fmt.Sprintf("%.1f MB", podTotalMemory)
+			newInfo.CPU = fmt.Sprintf("%.1f m", podTotalCPU)
+		}
+
+		podInfo = append(podInfo, newInfo)
+	}
+
+	return &types.Node{
+		Name:             node.Name,
+		IsConnected:      isConnected(node),
+		IsReady:          isReady(node),
+		IsPrimaryNode:    isPrimary(node),
+		CanDelete:        node.Spec.Unschedulable && !isConnected(node),
+		KubeletVersion:   node.Status.NodeInfo.KubeletVersion,
+		KubeProxyVersion: node.Status.NodeInfo.KubeProxyVersion,
+		OperatingSystem:  node.Status.NodeInfo.OperatingSystem,
+		KernelVersion:    node.Status.NodeInfo.KernelVersion,
+		CPU:              cpuCapacity,
+		Memory:           memoryCapacity,
+		Pods:             podCapacity,
+		Labels:           nodeRolesFromLabels(node.Labels),
+		Conditions:       findNodeConditions(node.Status.Conditions),
+		PodList:          podInfo,
+	}, nil
+}
+
+// nodeRolesFromLabels parses a map of k8s node labels, and returns the roles of the node
+func nodeRolesFromLabels(labels map[string]string) []string {
+	toReturn := []string{}
+
+	numRolesStr, ok := labels[types.EMBEDDED_CLUSTER_ROLE_LABEL]
+	if !ok {
+		// the first node will not initially have a role label, but is a 'controller'
+		if val, ok := labels["node-role.kubernetes.io/control-plane"]; ok && val == "true" {
+			return []string{"controller"}
+		}
+		return nil
+	}
+	numRoles, err := strconv.Atoi(strings.TrimPrefix(numRolesStr, "total-"))
+	if err != nil {
+		fmt.Printf("failed to parse role label %q: %s", numRolesStr, err.Error())
+
+		return nil
+	}
+
+	for i := 0; i < numRoles; i++ {
+		roleLabel, ok := labels[fmt.Sprintf("%s-%d", types.EMBEDDED_CLUSTER_ROLE_LABEL, i)]
+		if !ok {
+			fmt.Printf("failed to find role label %d", i)
```

Comment:
should there be a `continue` after this?

Reply:
yes, good catch